api/src/main/java/io/grpc/ChoiceChannelCredentials.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Provides a list of {@link ChannelCredentials}, where any one may be used. The credentials are in
+ * preference order.
```

Comment:
If a user puts `InsecureChannelCredentials` as the first one, we will always use that one based on the logic of `ProtocolNegotiators.from` so the list is meaningless. Although you may consider that as a user error should we check and not accept such input?

Reply:
> If they put TlsChannelCredentials as the first one, we will always use that one based on the logic in `ProtocolNegotiators.from`. I don't see that as meaningless though.

For `TlsChannelCredentials` there can be an error (due to `incomprehensible` being non-empty) so it can go to the next one. But for insecure no error is possible so it will always be picked. But I agree with you there is no need to enforce any input here based on logic that exists elsewhere.
> 
> I agree that users can configure ChoiceChannelCredentials in a way such that there's no point in ChoiceChannelCredentials, but that is context-dependent and we shouldn't be trying to police that. For all we know some developer will use some user-provided configuration to generate this list and not optimize the result. Note that we don't prohibit a single entry in the list, and I think that's fair. (We could potentially optimize the single-entry case in create(), but that seems more likely to cause bugs than help anyone.)